go/cmd/dolt/commands/cherry-pick.go

```diff
@@ -0,0 +1,466 @@
+		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+	}
+	workingRoot, err := dEnv.WorkingRoot(ctx)
+	if err != nil {
+		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+	}
+	headHash, err := headRoot.HashOf()
+	if err != nil {
+		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+	}
+	workingHash, err := workingRoot.HashOf()
+	if err != nil {
+		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+	}
+	if !headHash.Equal(workingHash) {
+		return HandleVErrAndExitCode(errhand.BuildDError("You must commit any changes before using cherry-pick.").SetPrintUsage().Build(), usage)
+	}
+
+	cherryStr := apr.Arg(0)
+	if len(cherryStr) == 0 {
+		verr := errhand.BuildDError("error: cannot cherry-pick empty string").Build()
+		return HandleVErrAndExitCode(verr, usage)
+	}
+
+	newWorkingRoot, commitMsg, err := CherryPick(ctx, dEnv, workingRoot, cherryStr)
+	if err != nil {
+		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+	}
+
+	workingHash, err = newWorkingRoot.HashOf()
+	if err != nil {
+		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+	}
+
+	if headHash.Equal(workingHash) {
+		cli.Println("No changes were made.")
+		return 0
+	}
+
+	err = dEnv.UpdateWorkingRoot(ctx, newWorkingRoot)
+	if err != nil {
+		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+	}
+	res := AddCmd{}.Exec(ctx, "add", []string{"-A"}, dEnv)
+	if res != 0 {
+		return res
+	}
+
+	// Pass in the final parameters for the author string.
+	commitParams := []string{"-m", commitMsg}
+	authorStr, ok := apr.GetValue(cli.AuthorParam)
+	if ok {
+		commitParams = append(commitParams, "--author", authorStr)
+	}
+
+	return CommitCmd{}.Exec(ctx, "commit", commitParams, dEnv)
+}
+
+// CherryPick returns updated Root value for current HEAD and commit message of cherry-picked commit.
+func CherryPick(ctx context.Context, dEnv *env.DoltEnv, wRoot *doltdb.RootValue, cherryStr string) (*doltdb.RootValue, string, error) {
+	opts := editor.Options{Deaf: dEnv.BulkDbEaFactory(), Tempdir: dEnv.TempTableFilesDir()}
+
+	cherrySpec, err := doltdb.NewCommitSpec(cherryStr)
+	if err != nil {
+		return nil, "", err
+	}
+	cherryCommit, err := dEnv.DoltDB.Resolve(ctx, cherrySpec, dEnv.RepoStateReader().CWBHeadRef())
+	if err != nil {
+		return nil, "", err
+	}
+
+	cherryCM, err := cherryCommit.GetCommitMeta(ctx)
+	if err != nil {
+		return nil, "", err
+	}
+	commitMsg := cherryCM.Description
+
+	updatedRoot, err := cherryPickACommit(ctx, dEnv.DoltDB, wRoot, cherryCommit, opts)
+	if err != nil {
+		return nil, "", err
+	}
+
+	return updatedRoot, commitMsg, nil
+}
+
+// cherryPickACommit takes working root and cherry-pick commit and returns updated root with changes from cherry-pick commit applied.
+func cherryPickACommit(ctx context.Context, ddb *doltdb.DoltDB, headRoot *doltdb.RootValue, cherryCM *doltdb.Commit, opts editor.Options) (*doltdb.RootValue, error) {
```

Comment:
we've got a couple of very similar function names here... `cherryPick` and `cherryPickACommit`. Would be good to rename this second one to something more self-descriptive that helps readers easily understand the difference between the two methods.